library/ssl_tls.c

```diff
@@ -5661,17 +5670,20 @@ int mbedtls_ssl_setup( mbedtls_ssl_context *ssl,
                        const mbedtls_ssl_config *conf )
 {
     int ret;
-    const size_t len = MBEDTLS_SSL_BUFFER_LEN;
+    const size_t in_len = MBEDTLS_SSL_IN_BUFFER_LEN;
+    const size_t out_len = MBEDTLS_SSL_OUT_BUFFER_LEN;
 
     ssl->conf = conf;
+    ssl->in_buf = NULL;
+    ssl->out_buf = NULL;
 
     /*
      * Prepare base structures
      */
-    if( ( ssl-> in_buf = mbedtls_calloc( 1, len ) ) == NULL ||
-        ( ssl->out_buf = mbedtls_calloc( 1, len ) ) == NULL )
+    if( ( ssl->in_buf = mbedtls_calloc( 1, in_len ) ) == NULL ||
```

Comment:
This is a very poor construct. We only handle the error on an OR, and always free the `in_buf`. What if the `in_buf` failed? This is a security bug - CWE-590.

Each allocation should be tested and handled individually.

Reply:
ssl->in_buf & ssl->out_buf are both initialized to NULL at line 5677, so the only possibilities are free(NULL) or free(freshly allocated in_buf).

Reply:
(I've removed this construct as part of revisions in #536 )